File: xls2csv.go
package main

import (
	"encoding/csv"
	"fmt"
	"os"
	"strings"

	"github.com/xuri/excelize/v2"
)

func cleanText(text string) string {
	// Replace any combination of \r\n or \n with a space
	return strings.ReplaceAll(strings.ReplaceAll(text, "\r\n", " "), "\n", " ")
}

func excelToCSV(filePath string) error {
	// Open Excel file
	f, err := excelize.OpenFile(filePath)
	if err != nil {
		return fmt.Errorf("error opening Excel file: %w", err)
	}
	defer f.Close()

	// Get all sheet names
	sheetNames := f.GetSheetList()

	// Process each sheet
	for _, sheetName := range sheetNames {
		// Read all rows from the sheet
		rows, err := f.GetRows(sheetName)
		if err != nil {
			return fmt.Errorf("error reading sheet %s: %w", sheetName, err)
		}

		// Clean data: replace line breaks with spaces
		cleanRows := make([][]string, len(rows))
		for i, row := range rows {
			cleanRows[i] = make([]string, len(row))
			for j, cell := range row {
				cleanRows[i][j] = cleanText(cell)
			}
		}

		// Create CSV file
		csvFileName := sheetName + ".csv"
		csvFile, err := os.Create(csvFileName)
		if err != nil {
			return fmt.Errorf("error creating CSV file %s: %w", csvFileName, err)
		}
		defer csvFile.Close()

		// Create CSV writer
		writer := csv.NewWriter(csvFile)
		defer writer.Flush()

		// Write all rows to CSV
		for _, row := range cleanRows {
			if err := writer.Write(row); err != nil {
				return fmt.Errorf("error writing to CSV file %s: %w", csvFileName, err)
			}
		}

		fmt.Printf("Worksheet '%s' has been exported to '%s'\n", sheetName, csvFileName)
	}

	return nil
}

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <excel_file>\n", os.Args[0])
		os.Exit(1)
	}

	filePath := os.Args[1]
	// Check if file exists
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		fmt.Fprintf(os.Stderr, "Error: File '%s' not found\n", filePath)
		os.Exit(1)
	}

	if err := excelToCSV(filePath); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}

File: xls2csv.py
import pandas as pd
import sys
import os

def excel_to_csv(file_path):
    """
    Convert Excel file to CSV files, one per worksheet.
    Replaces line breaks with spaces in cell contents.
    
    Args:
        file_path (str): Path to the Excel file
    """
    # Check if file exists
    if not os.path.exists(file_path):
        print(f"Error: File '{file_path}' not found")
        sys.exit(1)
        
    # Read the Excel file
    excel_data = pd.ExcelFile(file_path)
    
    # Iterate through each sheet and clean data
    for sheet_name in excel_data.sheet_names:
        df = pd.read_excel(excel_data, sheet_name=sheet_name)
        
        # Replace line breaks with space
        df = df.replace({r'[\r\n]+': ' '}, regex=True)
        
        # Export the cleaned data to CSV
        csv_file_path = f"{sheet_name}.csv"
        df.to_csv(csv_file_path, index=False, encoding='utf-8')
        print(f"Worksheet '{sheet_name}' has been exported to '{csv_file_path}'.")

def main():
    if len(sys.argv) != 2:
        print("Usage: python xls2csv.py <excel_file>")
        sys.exit(1)
    
    file_path = sys.argv[1]
    excel_to_csv(file_path)

if __name__ == "__main__":
    main()

